dd-sdk-android/src/test/kotlin/com/datadog/android/log/internal/net/CallbackNetworkInfoProviderTest.kt

```diff
@@ -0,0 +1,182 @@
+    lateinit var mockNetwork: Network
+    @Mock
+    lateinit var mockCapabilities: NetworkCapabilities
+
+    @BeforeEach
+    fun `set up`() {
+        whenever(mockCapabilities.hasTransport(any())) doReturn false
+
+        testedProvider = CallbackNetworkInfoProvider()
+    }
+
+    @Test
+    fun `initial state is not connected`() {
+        val networkInfo = testedProvider.getLatestNetworkInfo()
+
+        assertThat(networkInfo)
+            .hasConnectivity(NetworkInfo.Connectivity.NETWORK_NOT_CONNECTED)
+            .hasCarrierName(null)
+            .hasCarrierId(-1)
+    }
+
+    @Test
+    fun `connected to wifi`() {
+        whenever(mockCapabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI)) doReturn true
+
+        testedProvider.onCapabilitiesChanged(mockNetwork, mockCapabilities)
+        val networkInfo = testedProvider.getLatestNetworkInfo()
+
+        assertThat(networkInfo)
+            .hasConnectivity(NetworkInfo.Connectivity.NETWORK_WIFI)
+            .hasCarrierName(null)
+            .hasCarrierId(-1)
+    }
+
+    @Test
+    fun `connected to wifi aware`() {
+        whenever(mockCapabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI_AWARE))
+            .doReturn(true)
+
+        testedProvider.onCapabilitiesChanged(mockNetwork, mockCapabilities)
+        val networkInfo = testedProvider.getLatestNetworkInfo()
+
+        assertThat(networkInfo)
+            .hasConnectivity(NetworkInfo.Connectivity.NETWORK_OTHER)
+            .hasCarrierName(null)
+            .hasCarrierId(-1)
+    }
+
+    @Test
+    fun `connected to cellular`() {
+        whenever(mockCapabilities.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR))
+            .doReturn(true)
+
+        testedProvider.onCapabilitiesChanged(mockNetwork, mockCapabilities)
+        val networkInfo = testedProvider.getLatestNetworkInfo()
+
+        assertThat(networkInfo)
+            .hasConnectivity(NetworkInfo.Connectivity.NETWORK_CELLULAR)
+            .hasCarrierName(null)
+            .hasCarrierId(-1)
+    }
+
+    @Test
+    fun `connected to ethernet`() {
+        whenever(mockCapabilities.hasTransport(NetworkCapabilities.TRANSPORT_ETHERNET))
+            .doReturn(true)
+
+        testedProvider.onCapabilitiesChanged(mockNetwork, mockCapabilities)
+        val networkInfo = testedProvider.getLatestNetworkInfo()
+
+        assertThat(networkInfo)
+            .hasConnectivity(NetworkInfo.Connectivity.NETWORK_ETHERNET)
+            .hasCarrierName(null)
+            .hasCarrierId(-1)
+    }
+
+    @Test
+    fun `connected to VPN`() {
+        whenever(mockCapabilities.hasTransport(NetworkCapabilities.TRANSPORT_VPN)) doReturn true
+
+        testedProvider.onCapabilitiesChanged(mockNetwork, mockCapabilities)
+        val networkInfo = testedProvider.getLatestNetworkInfo()
+
+        assertThat(networkInfo)
+            .hasConnectivity(NetworkInfo.Connectivity.NETWORK_OTHER)
+            .hasCarrierName(null)
+            .hasCarrierId(-1)
```

Comment:
carrier id is always -1 ? Actually Hugo said something that he could not see the carrier id on the emulator. I told him it might be because the emulator doesn't have a real SIM Card.

Reply:
For Hugo's trouble, that's normal. Using the new methodology to get the network, we don't have the carrier information anymore, and since we don't have a context instance in the Callback class we can't retrieve it.

Reply:
We do have some Kbps download capabilities but it won't give the exact information. (Same value could be a powerfull Edge or a really bad 4G)